### ISSUE TITLE
Add support for reporting a problem with custom rule metadata

### ISSUE THREAD

**Issue by chloerice** [stylelint/stylelint#6513]
## tl;dr: 

I want to be able to `report` a `Problem` with custom metadata.

---

`customMessages` are set on `result.stylelint` by the `lintPostcssResult` method if the rule's config has a `message` in its secondary options. Similarly `ruleMetadata` is set on `result.stylelint` by `lintPostcssResult`, but instead of supporting custom `meta` in secondary options, the metadata is set with the rule function's `meta` property if it exists.

https://github.com/stylelint/stylelint/blob/d91bb5b779ce6ca69058549d58195a9e82b7c8d1/lib/lintPostcssResult.js#L100-L101

Currently, the `report` utility sets the warning message using either the `message` set on the problem reported or the message set on `result.stylelint.customMessages` if it exists. But it does not support overriding/setting the `result.stylelint.ruleMetadata` with custom metadata set on the problem reported.

https://github.com/stylelint/stylelint/blob/35c0cfa244f725a3548e8ba884edf70c2bc67732/lib/utils/report.js#L17-L26

### What solution would you like to see?

We can report problems with custom metadata by hard coding it onto `result.stylelint.ruleMetadata` before invoking the `report` utility with the problem. 

But I think it would be best for the `report` API to support custom metadata, just as it supports a custom `message` and `severity`, so that consumers don't have to reach into stylelint internal APIs when reporting problems.

```js
(root, result) => {

  for (const rule of rules) {
    const {ruleName, ruleSettings, fix, meta, customMessage, severity}

    stylelint.utils.checkAgainstRule({ruleName, ruleSettings, root, result, fix}, (warning) => {    
      stylelint.utils.report({
        result,
        ruleName,
        meta,
        severity,
        message: customMessage || warning.text,
        node: warning.node,
      })
    })
  });
 }
```

**Comment by ybiquitous** [comment]
@chloerice Thanks for using the template. It sounds interesting. What kind of background or motivation do you have for the suggestion?

**Comment by chloerice** [comment]
Hey @ybiquitous! The plugin my team is building is configured with a multi-dimensional rule config as its primary option, which consists of stylelint built-in and custom rule configs grouped by category. 

For example:

```js
rules: {
  'stylelint-polaris/coverage':   {
    colors: [{
      'color-named': 'never',
      'color-no-hex': true,
    }, {message: 'Please use a Polaris color token'}],
    conventions: [{
      'stylelint-polaris/media-queries-allowed-list': {
        allowedMediaTypes: ['print', 'screen'],
      },
    }],
  },
},
```

The Stylelint rules configured in each category are reported with the category's name (e.g., "colors") concatenated to the plugin's `ruleName` (e.g., "stylelint-polaris/coverage/_**colors**_") using the `report` utility in the callback returned from `checkAgainstRule`. Since these category rules are reported within this plugin instead of being defined and configured as separate plugins in the stylelint config, only the plugin rule function is called by the `lintPostcssResult` method, so no `ruleMetadata` can be automatically set for the dynamic rules.

In order to report these dynamic rules with their metadata, we can hard code the metadata onto `result.stylelint.ruleMetadata` directly before invoking the `report` utility with the problem. But we'd prefer to use the `report` API itself to set the custom `meta` for the warning like we can for a custom `message` and custom `severity`. 

```js
(root, result) => {
  for (const rule of stylelintRules) {
    const {categoryRuleName, stylelintRuleName, ruleSettings, fix, meta, message, severity} = rule;

    stylelint.utils.checkAgainstRule(
      {ruleName: stylelintRuleName, ruleSettings, root, result, fix},
      (warning) => {
        // Hard-code method
        result.stylelint.ruleMetadata[categoryRuleName] = meta;
    
        stylelint.utils.report({
          result,
          ruleName: categoryRuleName,
          // Preferred method
          // meta,
          severity,
          message: message || warning.text,
          node: warning.node,
        });
      }
    });
  }
}
```

**Comment by chloerice** [comment]
Thanks for following up @ybiquitous! I dug into the problem with @aaronccasanova yesterday to look for an existing convention to extend or apply changes to the `result`. We looking through the related source code and the current public API exposed. We're feel more confident in our current approach of setting the metadata on the result directly since the property is part of the `LinterResult` returned from `stylelint.lint`, so I'll close this issue.
